### PR TITLE
Format mxids when disambiguation needed

### PR DESCRIPTION
**Pull request by SimonBrandner**
### Description

Fixes [#16897](https://github.com/vector-im/element-web/issues/16897)
**Requires https://github.com/matrix-org/matrix-js-sdk/pull/1730**

### Dark

No display name:
Screenshot_20210420_120339

Disambiguated + flair: 
Screenshot_20210421_180130

### Light

Disambiguated + flair: 
Screenshot_20210421_180249

### CODE REVIEW

**Comment by t3chguy** [comment]
Does this *always* disambiguate rather than only when needed?

**Comment by SimonBrandner** [comment]
Yep, I misunderstood the issue... I'll rework this. But at least there is something to play with :D

**Comment by t3chguy** [comment]
it looks sleek

**Comment by t3chguy** [comment]
Could you please add screenshots exercising Flair and a no displayname user please

File: res/css/views/messages/_SenderProfile.scss
Comment (t3chguy) [inline res/css/views/messages/_SenderProfile.scss:20]: This looks to be unsupported in Safari
https://caniuse.com/flexbox-gap

File: src/components/views/messages/SenderProfile.js
Comment (t3chguy) [inline src/components/views/messages/SenderProfile.js:105]: with TS we could clean these up using `?.` to be like `{ mxEvent.sender?.userId || mxEvent.getSender() || "" }` or similar which IMO would be far more readable

**Review by t3chguy — CHANGES_REQUESTED** [review]
Code looks fine

Could benefit from a TS conversion if you fancy it

CSS unsupported in Safari is a blocker, unless PostCSS does some magic on it, let me know if it does.

File: src/components/views/messages/SenderProfile.js
Comment (SimonBrandner) [reply src/components/views/messages/SenderProfile.js:105]: Good point - will convert

File: src/components/views/messages/SenderProfile.tsx
Comment (t3chguy) [inline src/components/views/messages/SenderProfile.tsx:39]:
```suggestion
    private unmounted: boolean;
```

**Review by t3chguy — APPROVED** [review]
Code LGTM

File: src/components/views/messages/SenderProfile.tsx
Comment (SimonBrandner) [reply src/components/views/messages/SenderProfile.tsx]: Why do I always forget about this! :D

**Comment by niquewoodhouse** [comment]
Really like this change!

Tried to review but couldn't find a room with disambiguation happening 🤔 - all the rooms I've seen it in before just didn't have it, am I missing something?

image

Judging from the screenshot (so its a bit harder to give solid feedback)

- Colour: could you please confirm the colour of the disambiguation?  Is it the same as the timestamp?

- The [ ] has made the baseline jump so the username and display name characters look like they have different baselines, which makes it harder to read.   For example, on Twitter the `@` and the `name` have the same baseline, making it easier to scan across and understand the order.  And on Discord, the timestamp has the same baseline as the username.   Every small misalignment in the timeline can distract the eye and make it slightly harder to read it.  This kind of leads me into the next point...

image

- I don't personally get the [ ].  People can create accounts and use any Matrix client without knowing what Matrix is, let alone seeing/remembering the matrix logo so the connection feels tenuous, confusing and, in the product of Element, inconsistent.  For example, I can't see it on the user profile panel, but I can see it here...why?  What does it all mean? 🤔

image
  
If people want to use the [ ] in the product, I'd personally suggest it's a separate issue and looks at it consistently across the product.  I'd still personally advocate for it not being there but at least the consistency point would be irrelevant.  Maybe I'm missing something really obvious because I don't necessarily know Matrix as well as others but it doesn't feel to me like the right thing to arbitrarily do here.  

- What happens if I click on the [@jewjijwi:homeserver.com]? Is it the same as if I click on the name, or the avatar?  Or nothing?  

Thanks

**Comment by turt2live** [comment]
(the easiest way to cause disambiguation is to set up a GitHub notification bot or RSS bot through the integration manager)

**Comment by ShadowJonathan** [comment]
> If people want to use the [ ] in the product, I'd personally suggest it's a separate issue and looks at it consistently across the product. I'd still personally advocate for it not being there but at least the consistency point would be irrelevant. Maybe I'm missing something really obvious because I don't necessarily know Matrix as well as others but it doesn't feel to me like the right thing to arbitrarily do here.

This indeed does sound like a Product issue, I think that for now it'll be good to keep consistency and keep `()` parentheses around the disambiguation tags. I'll create another issue in element-web (for Product) that asks this question.

**Comment by turt2live** [comment]
I'm not convinced it needs to be deferred work here... we can just fix it in this PR, surely?

**Comment by SimonBrandner** [comment]
> I'm not convinced it needs to be deferred work here... we can just fix it in this PR, surely?

Yeah, definitely

**Comment by SimonBrandner** [comment]
OOps

**Comment by ShadowJonathan** [comment]
> I'm not convinced it needs to be deferred work here... we can just fix it in this PR, surely?

No i meant; Change it back to `()` here, and i'll make an issue in element-web discussing if using `[]` instead is okay.

**Comment by turt2live** [comment]
I'd rather we discuss it here given it's entirely within the context of this change. Also from the looks of it, the suggestion seems to be heading towards no brackets, not square vs round.

**Comment by ShadowJonathan** [comment]
> What happens if I click on the [@jewjijwi:homeserver.com]? Is it the same as if I click on the name, or the avatar? Or nothing?

The name, AFAIK, so i'll insert a mention into the current composer.

**Comment by ShadowJonathan** [comment]
> Tried to review but couldn't find a room with disambiguation happening :thinking: - all the rooms I've seen it in before just didn't have it, am I missing something?

I'm in the `#element-dev` room with an alt, so you can scroll up (or ping me to make some messages) and observe the changes.

**Review by t3chguy — CHANGES_REQUESTED** [review]
> Could you please add screenshots exercising Flair and a no displayname user please

**Comment by SimonBrandner** [comment]
> Could you please add screenshots exercising Flair and a no displayname user please

It's on my list of things to do - sorry for the delay

**Comment by t3chguy** [comment]
(I was noting it as a request so it doesn't get lost and someone merge without them is all, no rush)

**Comment by SimonBrandner** [comment]
@niquewoodhouse, I've updated the PR description - hopefully, it matches expectations. (I am not 100% happy about the flair but I am not sure how to improve it :( )

+ Color should now match the timestamp
+ `[]` was removed
+ Baseline should be the same now

**Comment by turt2live** [comment]
@SimonBrandner looks like you have a legitimate end-to-end test failure to play around with. Almost certainly it's because a selector changed, so the test script can no longer accurately find the right tile to pick apart.

**Comment by SimonBrandner** [comment]
It looked a little weird. Thanks for pointing me in the right direction

**Review by t3chguy — APPROVED** [review]
Code LGTM, thanks

**Comment by ara4n** [comment]
folks, is there a reason not to merge this? this would have helped avoid dramas during the freenode->libera migration where people were DMing pages of rage at me about the evils of impersonation attacks on Matrix...

**Comment by t3chguy** [comment]
> folks, is there a reason not to merge this? this would have helped avoid dramas during the freenode->libera migration where people were DMing pages of rage at me about the evils of impersonation attacks on Matrix...

Its awaiting Design review, @nadonomy can this get some attention?

**Comment by nadonomy** [comment]
@t3chguy @SimonBrandner from the screenshots it looks like we're using some different typography for the mxid.

What's the best way to test this in the [netlify build](https://deploy-preview-5880--matrix-react-sdk.netlify.app)? Join a room and observe mxid's being printed if people share a display name?

**Comment by t3chguy** [comment]
Yup that's right, maybe join #test:matrix.org and set your /myroomnick to that of someone in the recent timeline then send a message

**Comment by ShadowJonathan** [comment]
> What's the best way to test this in the netlify build? Join a room and observe mxid's being printed if people share a display name?

Yes, you can join #matrix-spec:matrix.org, for convenience, i just changed my alternate account's name to my main one, so you can observe both names.

**Comment by SimonBrandner** [comment]
> @t3chguy @SimonBrandner from the screenshots it looks like we're using some different typography for the mxid.

Yep, that was the point. Users had issues understanding what the mxid meant, so we were hoping that if we used a different font and color, it would be more understandable. You can read the original issue that this PR fixes for more detail: https://github.com/vector-im/element-web/issues/16897

**Comment by ShadowJonathan** [comment]
The typography change was my decision partially also, I thought that a monospaced font would be better representative of a "user ID", something closer to "technical", as (imo) monospace == technical.

If that's going to be an issue to grab from the existing typography templates, i'll personally be okay with changing it to a "known" font right now, and think about which monospace font to add to the element design templates later, and then to apply to this.

**Comment by SimonBrandner** [comment]
The preview is probably broken since I didn't update the js-sdk branch. I'll do that when I get home - 6 hours from now

**Comment by ShadowJonathan** [comment]
New js-sdk PR: https://github.com/matrix-org/matrix-js-sdk/pull/1730

**Comment by ShadowJonathan** [comment]
I'm not exactly sure if the brightness value of that last screenshot would be exactly what I was aiming for (a de-emphasised piece of information that can - if needed - be glazed over), the brightness here is as high as the actual username.

Wouldn't it work to lower it, to either some brightness like the state-change motif above ("[...] joined the room"), or somewhere in between? Sorry to junction this quick design suggestion.

**Comment by SimonBrandner** [comment]
@nadonomy, thanks for the review! I've updated the styling based on your suggestions

**Comment by ShadowJonathan** [comment]
Poking at devtools a little and picking `#81848a`, i get the following;

image

@nadonomy what do you think about this?

**Comment by SimonBrandner** [comment]
> Poking at devtools a little and picking `#81848a`, i get the following;
> 
> image
> 
> @nadonomy what do you think about this?

Personally, I like this one better

**Comment by nadonomy** [comment]
Gotcha. Deferring it in the hierarchy is a great idea/suggestion. However, we shouldn't insert new magic values into the codebase. It creates debt (at one point we had >100 different shades of grey) and we're working to unify all the styles in the app.

It'd be better to use an existing style, to ease unifying them.

From inspecting, can we match what we do for the join/leave styles pls? It just inherits body colour (so in light: `#2e2f32`, dark: `#FFFFFF`) with `opacity: 0.5` and visually will look like the screenshots posted.

**Comment by ShadowJonathan** [comment]
> It'd be better to use an existing style, to ease unifying them.

Of course, that's what I meant to intend, though I didn't write that down explicitly. I saw somewhere else there's a color style Figma document, and efforts to unify the colors to that document, I eyeballed the color because I wanted to convey the meaning, I'll leave it up to you and Simon to find the right one, but yes, I meant to "defer it in the hierarchy", I just couldn't find those words.

**Comment by SimonBrandner** [comment]
@nadonomy, I've updated the styling, so now it should look like join/leave (and other `.mx_TextualEvent`):

Screenshot_20210610_153213
Screenshot_20210610_153159

**Review by nadonomy — APPROVED** [review]
Great! Thanks all. :)

On these screenshots the font weight looks like it might be a touch too heavy, but hard to tell without using it in practise across multiple screens, resolutions, aliasing settings etc.

Approving, but may tweak the visuals after more testing!

**Comment by SimonBrandner** [comment]
Thanks for all the feedback

> Approving, but may tweak the visuals after more testing!

Sure, let me know if you have any further suggestions after using this in the wild

**Comment by SimonBrandner** [comment]
(re-requested a code review as I've made a few changes)

File: src/components/views/messages/SenderProfile.tsx
Comment (t3chguy) [inline src/components/views/messages/SenderProfile.tsx:125]: can we just inline this into where we render it? Think it'll be cleaner

Comment (SimonBrandner) [reply src/components/views/messages/SenderProfile.tsx:125]: This actually seemed a bit clearer to me this way but sure I'll change it when I get home :D

**Pull request merged**
